include/ringmesh/basic/attributes.h

```diff
@@ -0,0 +1,1024 @@
+         */
+        ~AttributeStore();
+
+        void set_store( Store* store )
+        {
+            store_.reset( store );
+        }
+
+        const Store& get_store() const
+        {
+            return *store_;
+        }
+
+        /**
+         * \brief Resizes this AttributeStore
+         * \param[in] new_size new number of items
+         */
+        void resize( index_t new_size )
+        {
+            store_->resize( new_size );
+        }
+
+        /**
+         * \brief Resizes this AttributeStore to 0.
+         */
+        void clear()
+        {
+            store_->clear();
+        }
+
+        /**
+         * \brief Applies a permutation to the stored attributes.
+         * \details Applying a permutation to the data is equivalent
+         *  to:
+         * \code
+         * for(i=0; i<permutation.size(); i++) {
+         *    data2[i] = data[permutation[i]]
+         * }
+         * data = data2 ;
+         * \endcode
+         * But it is done in-place.
+         * \param[in] permutation the permutation.
+         *  It is temporarily changed during execution of the
+         *  function, but identical to the input on exit.
+         * \note This function uses memcpy(). If required, it
+         *  can be overloaded in derived classes.
+         */
+        void apply_permutation( const std::vector< index_t >& permutation )
+        {
+            store_->apply_permutation( permutation );
+        }
+
+        /**
+         * \brief Compresses the stored attributes, by
+         *  applying an index mapping that fills-in the gaps.
+         * \details This is equivalent to:
+         * \code
+         * for(i=0; i<size(); i++) {
+         *    if(old2new[i] != index_t(-1)) {
+         *       data2[old2new[i]] = data[i];
+         *    }
+         * }
+         * data = data2 ;
+         * \endcode
+         * \param[in] old2new the index mapping to be applied.
+         * \pre old2new[i] <= i || old2new[i] == index_t(-1)
+         * \note This function uses memcpy(). If required, it
+         *  can be overloaded in derived classes.
+         */
+        void compress( const std::vector< index_t >& old2new )
+        {
+            store_->compress( old2new );
+        }
+
+        /**
+         * \brief Creates a new AttributeStore that is a carbon copy
+         *  of this AttributeStore.
+         * \details Only the data is copied.
+         */
+        std::unique_ptr< RINGMesh::AttributeStore > clone() const;
+
+        const void* element( index_t element ) const
+        {
+            return store_->element( element );
+        }
+
+        void* element( index_t element )
+        {
+            return store_->element( element );
+        }
+
+        bool elements_type_matches( const std::string& type_name ) const
+        {
+            return store_->elements_type_matches( type_name );
+        }
+
+    protected:
+        std::unique_ptr< Store > store_{ nullptr };

```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Make this member variable "private". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=cpp%3AS3656)